### ISSUE TITLE
Re-declaring let bindings that reference themselves causes segfault

### ISSUE THREAD

**Issue by theundeadmonk** [rspec/rspec-core#2339]
If i have the following let binding

`let(:var1) { create(:some_model) }`

And I proceed to subsequently define that binding again to reference itself
`let(:var1) { v = var1 }`

It results in a segfault

`Error Message:
6025 segmentation fault
[NOTE]
You may have encountered a bug in the Ruby interpreter or extension libraries.
Bug reports are welcome.
Don't forget to include the above Crash Report log file.
For details: http://www.ruby-lang.org/bugreport.http`

Not sure if this is a valid bug or I'm doing something wrong.

Thanks.

**Comment by myronmarston** [comment]
It is a bug, but it's a bug in Ruby, not a bug in RSpec.

It looks like you've hit this:

https://bugs.ruby-lang.org/issues/9505

You can follow up with ruby core to see if they will fix it (or upgrade to a version with a fix; the discussion in that issue suggests its fixed Ruby 2.2+).

**Comment by JoshCheek** [comment]
Can you provide a [SSCCE](http://sscce.org/)? I can think of two versions of what your code could looked like:

We'll call this one **A**

```ruby
require "rspec/autorun"
describe "outer" do
  let(:var1) { :something_about_active_record_i_think }
  let(:var1) { v = var1 }
  example { expect(var1).to eq :something_about_active_record_i_think }
end
```

We'll call this one **B**

```ruby
require "rspec/autorun"
describe "outer" do
  let(:var1) { :something_about_active_record_i_think }
  describe 'inner' do
    let(:var1) { v = var1 }
    example { expect(var1).to eq :something_about_active_record_i_think }
  end
end
```

Given **A**, I'd need to know more about what you're trying to do to provide useful feedback. The core relevant thing that's happening here is that your second let block is overriding your first one, the first one is gone. When it invokes `var1`, it's invoking itself, so I get a `SystemStackError` on Ruby 2.3.1

Given **B**, the inner `let` block may use `super()` to call the outer let block. Eg:

```ruby
require "rspec/autorun"
describe "outer" do
  let(:var1) { :something_about_active_record_i_think }
  describe 'inner' do
    let(:var1) { v = super() }
    example { expect(var1).to eq :something_about_active_record_i_think }
  end
end
```

Also note that the local variable assignment (`v = ...`) is doing nothing here.
